gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java

```diff
@@ -1916,23 +1915,25 @@ public WorldScheduler scheduler() {
 
     @Override
     public boolean mainThread() {
-        if (this.threadId == -1) {
-            return true;
-        } else {
-            long id = Thread.currentThread().getId();
-            return this.threadId == id || this.asyncAllowedThreadId == id;
-        }
+        return mainThread(false);
     }
 
-    public boolean hardMainThreadCheck() {
+    /**
+     * Check if current thread is this world's main thread
+     *
+     * @param allowForIOThread whether to allow async world io thread
+     * @return true if main thread, false if not
+     */
+    public boolean mainThread(boolean allowForIOThread) {
         if (this.threadId == -1) {
             return true;
         } else {
             long id = Thread.currentThread().getId();
-            return this.threadId == id;
+            return this.threadId == id || allowForIOThread && this.asyncAllowedThreadId == id;
```

Comment:
allowForIOThread && this.asyncAllowedThreadId == id need to be in ( ) ?

Reply:
&& binds stronger more than ||